### DISCUSSION TITLE
Custom Schema Plugin

### DISCUSSION THREAD

**Discussion by hand-dot** [pdfme/pdfme#208] [Ideas]
# Draft
- ## 1st Draft: https://github.com/pdfme/pdfme/discussions/208#discussioncomment-6643719

> This is just a proposal based on an idea, but what about creating a plugin mechanism and allowing your team to implement the placeholder functionality you desire through plugins? pdfme also supports barcodes and QR codes, but since they are unnecessary features for users who don't use them, I previously thought it might be good to make them into plugins. If this plugin mechanism can be implemented, you can define any schema type you like, and I think we can take advantage of being open-source.

context: https://github.com/pdfme/pdfme/discussions/208#discussioncomment-6610844

---

ref: Placeholder text support for mixing dynamic and static text content #205
ref: Use basic templating / placeholder substitution for text schemas #201

**Comment by hand-dot** [comment]
@peteward 
Could you please explain a bit more about why this feature is necessary, using more specific use case? I have three questions:

- Is it necessary to implement this in pdfme? Why not add the process to your app side?
- What will the UI for inputting with a form look like?
- I don't want to use a custom template engine. Have you considered using something like https://liquidjs.com/index.html?

**Reply by peteward** [reply]
Hey @hand-dot,

Our use-case allows ecommerce merchants to design their own PDF templates (using pdfme), and merge this with the data from orders placed by their customers. This is to create personalised cards to be printed and inserted as part of their orders.

In most of our designs there is a mixture of dynamic (variable-based) and static text, which needs to be entered together because the length of the dynamic content can impact the static positioning. 

Here are a few common, but simple examples:

<img width="431" alt="Screenshot 2023-07-31 at 15 14 14" src="https://github.com/pdfme/pdfme/assets/7068515/06c0d5d6-91bc-4854-8777-16c716cde2e4">
<img width="510" alt="Screenshot 2023-07-31 at 15 14 30" src="https://github.com/pdfme/pdfme/assets/7068515/5b73eb45-2464-49ee-8177-7893717d01f2">
<img width="406" alt="Screenshot 2023-07-31 at 15 14 55" src="https://github.com/pdfme/pdfme/assets/7068515/54f3eec1-80a6-4735-8ad6-3e5a7cdb609e">

There are other use-cases that are more complex, but **at the moment** our only requirement is text before or after the dynamic variable.


This mixture of static and dynamic text is very much part of the design process, so I think it makes most sense for this to be within pdfme. I'm not sure how we could do it within our app outside of pdfme as we are embedding the pdfme designer within our app:

![Screenshot 2023-07-31 at 15 20 33](https://github.com/pdfme/pdfme/assets/7068515/1711bb14-5ee1-4346-8ff4-b6a2d317419f)


Since all we wanted was basic variable substitution I have just hacked together this substitution manually (as it is our only immediate requirement). However, it makes sense to use a more robust but simple and lightweight templating engine.


In the future, I can definitely see use cases for more complex templating such as conditions and loops, possibly default/fallback values. 

I can also see complexity with the current structure if you wanted to have a text paragraph with 2 different variables in:

```
Hi {{name}}, thank you for buying {{product}}, we hope you love it!
```

This would require more thought, and it is not a short-term requirement for us, but will probably be asked for by our customers eventually...

**Comment by hand-dot** [comment]
Hi @peteward, Thanks for explaining.

I'm not sure if I understand correctly, but I feel that this feature is unnecessary.

Because
The values entered in the "Input Example" in the designer are used as "sampledata" of template when creating the template.
For fields entered as "Hello {{name}}!" in the "Input Example", 
I think all you need to do is add preprocessing before calling the generate function and replace the value of {{name}}.

If you just want to preview how it will turn out and see the appearance, it's okay to have the user temporarily rewrite the Input Example, or you can also set the changed value of the Template's sampledata with the designer's updateTemplate method.

I want to know why this workaround I explained doesn't work.

**Reply by peteward** [reply]
Hi @hand-dot,

Yes, that would certainly be achievable, but a less good user experience for our use case. Our customers will be needing this feature very often, probably most of the time.

I think templating/placeholders in general are not a great experience for non-techincal users, but I cannot think of a better way to implement this feature, therefore we want to make that experience as good as possible. The UI additions that I showed in the video feel more intuitive than users having to know to use the placeholders (and type them in correctly!).  You get to see sample data by toggling focus of the input box, along with rendered previews. 
Our internal team are using it this week so I will get some feedback on their experience soon.

An idea that we explored in the office today was:
* using text boxes without variables (i.e. just adding a text box to a pdf without anything dynamic)
* adding text boxes with multiple variables in

This got me thinking about the idea of decoupling text boxes from variables. Variables could then be simply defined with sample values and could be added to text boxes as required. 🤔  Again, this would be a big change though.

I completely understand though if none of this is either a priority for pdfme, or the direction you wish it to go. As you say, there are ways for the basic placeholder approach to be achieved if needed :)

**Reply by hand-dot** [reply]
@peteward 
Yes, I think this issue is important for the UX of your company's service users, and I respect that. However, by including this support in pdfme's text type schema, other pdfme users may find it complicated. This is something we would like to avoid since the concept of pdfme is that PDFs can be easily created.

This is just a proposal based on an idea, but what about creating a plugin mechanism and allowing your team to implement the placeholder functionality you desire through plugins? pdfme also supports barcodes and QR codes, but since they are unnecessary features for users who don't use them, I previously thought it might be good to make them into plugins. If this plugin mechanism can be implemented, you can define any schema type you like, and I think we can take advantage of being open-source.

What do you think?

**Reply by peteward** [reply]
Yes, I think that would be very sensible because otherwise you could end up with too much bloat of unwanted features and complexity for pdfme 👍

**Reply by hand-dot** [reply]
Cool, I'll change the title of this discussion to pdfme schema plugin.
Let's proceed discussion in here.

**Comment by hand-dot** [comment]
# Custom Schema 1st Draft

## Purpose

- To be able to process schema types that are not supported, by utilizing Custom Schema.
- The processing of previously supported schema types such as text, image, and various barcodes will be transferred to the newly created @pdfme/schema package.
- Users can develop their own Custom Schemas with reference to @pdfme/schema, allowing them to address various issues.

## Modules

Pdfme provides not only PDF rendering but also multiple modules such as Designer's WYSIWYG, Property Editor, Form, Viewer, etc. Users can achieve the goal by adding Custom Schema processing to the required modules.

<details>
<summary>@pdfme/common</summary>

  - Custom Validator
      - Enable Custom Schema-compatible validation in zod's validation called in the [ui constructor](https://github.com/pdfme/pdfme/blob/b5c176b4981186e1d91811c04bdda2a350d17487/packages/ui/src/class.ts#L80), [inside ui's updateTemplate](https://github.com/pdfme/pdfme/blob/b5c176b4981186e1d91811c04bdda2a350d17487/packages/ui/src/class.ts#L115), and [preprocessing of generator](https://github.com/pdfme/pdfme/blob/b5c176b4981186e1d91811c04bdda2a350d17487/packages/generator/src/generate.ts#L46).
</details>

<details>
<summary>@pdfme/generator</summary>

- PDF renderer
  - Extend [drawInputByTemplateSchema](https://github.com/pdfme/pdfme/blob/b5c176b4981186e1d91811c04bdda2a350d17487/packages/generator/src/helper.ts#L396) to execute Custom Schema PDF drawing processing.
</details>

<details>
<summary>@pdfme/ui</summary>

  - Designer
      - Custom Renderer
          - state: Write
              - Selecting a Field and clicking on it will result in write mode, allowing changes to the sample data. This mode runs the same code as Form.
                <img width="364" alt="Untitled" src="https://github.com/pdfme/pdfme/assets/24843808/458d7c1c-f1fd-4d1b-9a3b-8c1c468595f7">
          - state: Read
              - Anything other than the write mode above will be read mode. This mode runs the same code as Viewer.
                <img width="330" alt="Untitled (1)" src="https://github.com/pdfme/pdfme/assets/24843808/8e5f070a-11e3-4378-8cc9-ae389b4f1b16">
                <img width="337" alt="Untitled (2)" src="https://github.com/pdfme/pdfme/assets/24843808/b136032c-932d-4a00-a90f-c00413fc50e1">
      - Prop Editor
          - Users can define Custom Prop Editor, and also, since the Sample data Editor can write directly to the value in write mode, it is unnecessary. This will be removed as the user's settings become more complex.
            ![Untitled (3)](https://github.com/pdfme/pdfme/assets/24843808/09f37c41-340c-4ec9-9571-6df741f4c116)
  - Form
      - Custom Renderer
          - state:  Write
              - Runs the same code as Designer > Custom Renderer > state:  Write.
                    <img width="329" alt="Untitled (4)" src="https://github.com/pdfme/pdfme/assets/24843808/8c0ef902-d257-466f-a143-fd7b2b54141a"> 
  - Viewer
      - Custom Renderer
          - state:  Read
              - Runs the same code as Designer > Custom Renderer > state:  Read.
                    <img width="302" alt="CleanShot 2023-08-05 at 12 05 43@2x" src="https://github.com/pdfme/pdfme/assets/24843808/cf6f1046-3345-43e1-83dc-6301f63150aa">
</details>

## Interface

```js
const customSchemaForGenerator = {
  myschema: {
    renderer: ({ pdfDoc, page, schema, inputValue }) => {
      // TODO
    }
  }
};

const customSchemaForUI = {
  myschema: {
    renderer: ({ schemaProp, inputValue, onChange, isActive }) => { // for editor 
      // TODO
    },
    // renderer: ({ schemaProp, inputValue, onChange }) => { // form
      // TODO
    // },
    // renderer: ({ schemaProp, inputValue }) => { // viewer
      // TODO
    // },
    propEditor: ({ schemaProp, onChange }) => { // only for editor
      // TODO
    },
  }
};
```

- @pdfme/generator
    
    ```js
    generate({ template, inputs, options: { schema: customSchemaForGenerator } })
    ```
    
- @pdfme/ui
    
    ```js
    new Designer({ domContainer, template, options: { schema: customSchemaForUI } });
    ```
    

## Notes

- To keep the code small and clean in both generator and UI, process built-in and third-party schema types with a common interface.
- With this modification, pdfme only needs to handle the layout of each schema, and how to render it in PDF and HTML can be left to each schema.
- The schema part written in generator and UI should be separated into another package, reducing the amount of code in generator and UI.
- Although there are various barcodes currently, it is not considered necessary for all users, so they will be made opt-in. Built-in schema types will only include text and images.
- This will be a massive modification that overhauls the processing of generator and UI. The start date for implementation and the release date are still undetermined.

**Reply by peteward** [reply]
This is an excellent approach, I really like it!
I think the idea of removing the Sample Data from the sidebar is a positive change.

**Reply by hand-dot** [reply]
@peteward 
Thank you.

This feature is very ideal, but it requires a lot of modifications to be realized.
I am not sure yet when it can be released, as I will be working on the implementation on my days off and after my main job.

I will keep you posted if there are any updates.

**Reply by peteward** [reply]
Of course, no problem!

If there is anything we can do to help then please let us know.

**Comment by hand-dot** [comment]
I have decided to implement this feature. Updates will be made in  #213  going forward, so I will close this thread.
